server/src/main/java/org/elasticsearch/ingest/IngestService.java

```diff
@@ -340,222 +304,202 @@ public ClusterState execute(ClusterState currentState) {
-            Optional<StatsHolder> pipelineStats = Optional.ofNullable(statsHolderPerPipeline.get(pipeline.getId()));
-            try {
-                totalStats.preIngest();
-                pipelineStats.ifPresent(StatsHolder::preIngest);
-                String index = indexRequest.index();
-                String type = indexRequest.type();
-                String id = indexRequest.id();
-                String routing = indexRequest.routing();
-                Long version = indexRequest.version();
-                VersionType versionType = indexRequest.versionType();
-                Map<String, Object> sourceAsMap = indexRequest.sourceAsMap();
-                IngestDocument ingestDocument = new IngestDocument(index, type, id, routing, version, versionType, sourceAsMap);
-                pipeline.execute(ingestDocument);
-
-                Map<IngestDocument.MetaData, Object> metadataMap = ingestDocument.extractMetadata();
-                //it's fine to set all metadata fields all the time, as ingest document holds their starting values
-                //before ingestion, which might also get modified during ingestion.
-                indexRequest.index((String) metadataMap.get(IngestDocument.MetaData.INDEX));
-                indexRequest.type((String) metadataMap.get(IngestDocument.MetaData.TYPE));
-                indexRequest.id((String) metadataMap.get(IngestDocument.MetaData.ID));
-                indexRequest.routing((String) metadataMap.get(IngestDocument.MetaData.ROUTING));
-                indexRequest.version(((Number) metadataMap.get(IngestDocument.MetaData.VERSION)).longValue());
-                if (metadataMap.get(IngestDocument.MetaData.VERSION_TYPE) != null) {
-                    indexRequest.versionType(VersionType.fromString((String) metadataMap.get(IngestDocument.MetaData.VERSION_TYPE)));
-                }
-                indexRequest.source(ingestDocument.getSourceAndMetadata());
-            } catch (Exception e) {
-                totalStats.ingestFailed();
-                pipelineStats.ifPresent(StatsHolder::ingestFailed);
-                throw e;
-            } finally {
-                long ingestTimeInMillis = TimeUnit.NANOSECONDS.toMillis(System.nanoTime() - startTimeInNanos);
-                totalStats.postIngest(ingestTimeInMillis);
-                pipelineStats.ifPresent(statsHolder -> statsHolder.postIngest(ingestTimeInMillis));
-            }
+        if (changed) {
+            statsHolderPerPipeline = Collections.unmodifiableMap(newStatsPerPipeline);
         }
+    }
 
-        private Pipeline getPipeline(String pipelineId) {
-            Pipeline pipeline = store.get(pipelineId);
-            if (pipeline == null) {
-                throw new IllegalArgumentException("pipeline with id [" + pipelineId + "] does not exist");
+    private void innerExecute(IndexRequest indexRequest, Pipeline pipeline) throws Exception {
```

Comment:
maybe a more descriptive method name?

Reply:
@martijnvg any suggestions? :)

I agree that these names are kind of meh, but note: We have a bunch of other `inner*` methods in this class. In the end they're all kind of redundant since they're only used in one place (at least in production code that is) each => somewhat hard to find a better name for a method that effectively only exists to shorten the code of whatever method name that comes after the `inner` prefix.

Reply:
The inner method prefix was introduced to make the actual logic easier to unit test.

I was more referring to the execute part of this method, which is very generic :)
maybe `innerIngest()`?

Reply:
@martijnvg I like that (`innerIngest`), renaming :)